### PR TITLE
Explicitly derive alloca AS from DL

### PR DESCRIPTION
**Pull request by jpsamaroo**
Fixes issues around usage of the AMDGPU LLVM target by GPUCompiler

Supersedes #45544

### CODE REVIEW

**Comment by vtjnash** [comment]
still sounds good to me

**Comment by ianatol** [comment]
I don't think this is quite ready to merge --- I can recreate the `atomics` test failure locally on `macOS (arm64-apple-darwin21.5.0)`

**Comment by ianatol** [comment]
Well, I can do a bit better, here's a reproducer:
```
julia> (x, y) = (Complex{Int128}(10, 30), Complex{Int128}(20, 40))
(10 + 30im, 20 + 40im)

julia> mutable struct ARefxy{T}
                 @atomic x::T
                 y::T
                 ARefxy(x::T, y::T) where {T} = new{T}(x, y)
                 ARefxy{T}(x, y) where {T} = new{T}(x, y)
                 ARefxy{T}() where {T} = new{T}()
             end

julia> ar = ARefxy(x, y)

signal (11): Segmentation fault: 11
in expression starting at REPL[3]:1
_ZNK4llvm11Instruction9getModuleEv at /Users/ian/projects/45900/usr/lib/libLLVM.dylib (unknown line)
Allocations: 2870 (Pool: 2858; Big: 12); GC: 0
```

**Comment by vchuravy** [comment]
Can you run this in an assertion build or produce the LLVM IR?

**Comment by jpsamaroo** [comment]
Looks like `getModule` might have been called on an invalid `Instruction` (must have been null).

File: src/codegen.cpp [hunk 1]
```diff
@@ -3824,15 +3824,17 @@ static jl_cgval_t emit_call_specfun_other(jl_codectx_t &ctx, jl_method_instance_
...
     case jl_returninfo_t::Register:
     case jl_returninfo_t::Ghosts:
         break;
     case jl_returninfo_t::SRet:
-        result = emit_static_alloca(ctx, getAttributeAtIndex(returninfo.decl->getAttributes(), 1, Attribute::StructRet).getValueAsType());
+        ety = getAttributeAtIndex(returninfo.decl->getAttributes(), 1, Attribute::StructRet).getValueAsType();
+        result = emit_static_alloca(ctx, ety);
         assert(cast<PointerType>(result->getType())->hasSameElementTypeAs(cast<PointerType>(cft->getParamType(0))));
-        argvals[idx] = result;
+        argvals[idx] = ctx.builder.CreateAddrSpaceCast(result, ety->getPointerTo(0));
```
Comment (vchuravy) [inline src/codegen.cpp:3837]: Please add comments documenting the reason behind this cast.

**Comment by vchuravy** [comment]
Can we add some test to `test/llvmpasses`? The codegen changes are annoying to test for, but the finalgc/lategc, alloc-opt, and propagate-addrspaces should be testable.

**Comment by jpsamaroo** [comment]
I also plan to change sret so that instead of addrspacecasting the caller's alloca to AS 0, we set the sret pointer type to the correct AS in the callee signature.

**Comment by vchuravy** [comment]
> I also plan to change sret so that instead of addrspacecasting the caller's alloca to AS 0, we set the sret pointer type to the correct AS in the callee signature.

I was looking at the LLVM langref and it is not specified that sret needs to be an alloca. Maybe double check what clang produces?

**Comment by vtjnash** [comment]
`sret` is not an alloca: it can be anything that is a pointer. Optimization passes such as MemCpyOpt will make use of that to replace an alloca with the later memcpy destination for example

**Comment by jpsamaroo** [comment]
@vtjnash to confirm, I should go back to doing an `addrspacecast` of the sret slot?

**Comment by vtjnash** [comment]
Not necessarily. You are free to assert that some `sret` we emitted is required to be an alloca. It just is not true in general that an sret encountered "in the wild" will reference stack memory.

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45900#issuecomment-1180472951) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/655c6e0_vs_afe3838/report.html).

**Review by vchuravy — APPROVED** [review]

**Pull request merged**